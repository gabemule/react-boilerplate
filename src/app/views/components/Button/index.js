import React from 'react'

import { DsButton, DsParagraph, DsHeading } from '@gabemule/design-system'

const Button = () => {
  return (
    <div className={'Button'}>
      <DsHeading type="display">Buttons - Dynamic Theme Change</DsHeading>

      <br />
      <br />

      <DsParagraph type="small">Small</DsParagraph>

      <br />

      <DsButton size="small"> Primary </DsButton>
      <DsButton size="small" type="secondary">
        {' '}
        Secondary{' '}
      </DsButton>
      <DsButton size="small" type="warning">
        {' '}
        Warning{' '}
      </DsButton>
      <DsButton size="small" type="danger">
        {' '}
        Danger{' '}
      </DsButton>
      <DsButton disabled size="small">
        {' '}
        Disabled{' '}
      </DsButton>

      <br />
      <br />

      <DsParagraph type="small">Medium</DsParagraph>

      <br />

      <DsButton> Primary </DsButton>
      <DsButton type="secondary"> Secondary </DsButton>
      <DsButton type="warning"> Warning </DsButton>
      <DsButton type="danger"> Danger </DsButton>
      <DsButton disabled> Disabled </DsButton>

      <br />
      <br />

      <DsParagraph type="small">Large</DsParagraph>

      <br />

      <DsButton size="large"> Primary </DsButton>
      <DsButton size="large" type="secondary">
        {' '}
        Secondary{' '}
      </DsButton>
      <DsButton size="large" type="warning">
        {' '}
        Warning{' '}
      </DsButton>
      <DsButton size="large" type="danger">
        {' '}
        Danger{' '}
      </DsButton>
      <DsButton disabled size="large">
        {' '}
        Disabled{' '}
      </DsButton>
    </div>
  )
}

export default Button
