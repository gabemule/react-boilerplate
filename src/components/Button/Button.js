import React from 'react'
import PropTypes from 'prop-types'

import '@/components/Button/button.css'

const Button = ({ primary, backgroundColor, size, label, ...props }) => {
  const mode = primary
    ? 'storybook-button--primary'
    : 'storybook-button--secondary'
  return (
    <button
      className={['storybook-button', `storybook-button--${size}`, mode].join(
        ' '
      )}
      style={backgroundColor && { backgroundColor }}
      type="button"
      {...props}
    >
      {label}
    </button>
  )
}

Button.propTypes = {
  /**
   * What background color to use
   */
  backgroundColor: PropTypes.string,
  /**
   * Button contents
   */
  label: PropTypes.string.isRequired,
  /**
   * Optional click handler
   */
  onClick: PropTypes.func,
  /**
   * Is this the principal call to action on the page?
   */
  primary: PropTypes.bool,
  /**
   * How large should the button be?
   */
  size: PropTypes.oneOf(['small', 'medium', 'large'])
}

Button.defaultProps = {
  backgroundColor: null,
  onClick: undefined,
  primary: false,
  size: 'medium'
}

export default Button
