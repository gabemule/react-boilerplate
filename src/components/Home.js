import React from "react";
import MyImage from "../assets/dog.jpg";

const Home = () => {
    return (
        <div>
            <h1>Home</h1>
            <img src={MyImage} alt={'home image'} />
        </div>
    );
};

export default Home;
