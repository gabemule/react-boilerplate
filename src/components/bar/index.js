import Bar from './bar';

export default Bar;
