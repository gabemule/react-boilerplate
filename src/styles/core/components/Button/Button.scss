.ds-button {
  $self: &;

  font-family: $button__font--family;
  border: none;
  cursor: pointer;
  position: relative;
  text-align: center;
  overflow: hidden;
  box-sizing: border-box;

  &__text {
    display: flex;
    align-items: center;
    justify-content: center;
    font-family: $button__font--family;
    font-weight: $button__font--weight;
  }

  &__icon {
    margin-left: $spacing--nano;
  }

  // Size modifiers
  &--large {
    height: gettoken($button__large, "height");
    border-radius: gettoken($button__large, "border-radius");
    padding: gettoken($button__large, "padding");

    .ds-button__text {
      font-size: gettoken($button__large, "font-size");
    }
  }

  &--medium {
    height: gettoken($button__medium, "height");
    border-radius: gettoken($button__medium, "border-radius");
    padding: gettoken($button__medium, "padding");

    .ds-button__text {
      font-size: gettoken($button__medium, "font-size");
    }
  }

  &--small {
    height: gettoken($button__small, "height");
    border-radius: gettoken($button__small, "border-radius");
    padding: gettoken($button__small, "padding");

    .ds-button__text {
      font-size: gettoken($button__small, "font-size");
    }
  }

  // Type modifiers
  &--primary {
    background: gettoken($button__primary, "background");
    color: gettoken($button__primary, "color");
    border: gettoken($button__primary, "border");

    &:not(#{ $self }--disabled) {
      &:hover {
        background: gettoken($button__primary--hover, "background");
      }

      &:focus {
        outline-color: transparent;
        outline-style: solid;
        box-shadow: 0 0 0 3px var(--color__neutral--00),
          0 0 0 5px var(--color__neutral--90);
      }
    }

    &.ds-button--disabled {
      background: gettoken($button__primary--disabled, "background");
      color: gettoken($button__primary--disabled, "color");
    }
  }

  &--secondary {
    background: gettoken($button__secondary, "background");
    color: gettoken($button__secondary, "color");
    border: gettoken($button__secondary, "border");

    &:not(#{ $self }--disabled) {
      &::after {
        background: rgba(255, 255, 255, 0.3);
      }

      &:hover {
        background: gettoken($button__secondary--hover, "background");
        color: gettoken($button__secondary--hover, "color");

        & #{ $self }--icon {
          color: var(--button-text-color__secondary--hover);
          background: var(--button-background-color__secondary--hover);
        }

        & #{ $self }--text {
          color: var(--button-text-color__secondary--hover);
        }
      }
    }

    &.ds-button--disabled {
      background: gettoken($button__secondary--disabled, "background");
      border-color: gettoken($button__secondary--disabled, "color");
      color: gettoken($button__secondary--disabled, "color");
    }
  }

  &--warning {
    background: gettoken($button__warning, "background");
    color: gettoken($button__warning, "color");

    &:not(#{ $self }--disabled) {
      &:hover {
        background: gettoken($button__warning--hover, "background");
      }
    }

    &.ds-button--disabled {
      background: gettoken($button__warning--disabled, "background");
      color: gettoken($button__warning--disabled, "color");
    }
  }

  &--danger {
    background: gettoken($button__danger, "background");
    color: gettoken($button__danger, "color");

    &:not(#{ $self }--disabled) {
      &:hover {
        background: gettoken($button__danger--hover, "background");
      }
    }

    &.ds-button--disabled {
      background: gettoken($button__danger--disabled, "background");
      color: gettoken($button__danger--disabled, "color");
    }
  }

  &--disabled {
    outline: none;
    cursor: not-allowed;
    color: gettoken($button__disabled, "color");

    & #{ $self }__button-text {
      background: var(--color__neutral--10);
      color: gettoken($button__disabled, "color");
    }

    & #{ $self }__icon {
      background: gettoken($button__disabled, "background");
    }
  }
}