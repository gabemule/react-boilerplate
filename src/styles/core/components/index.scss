@import "Accordion/Accordion";
@import "Accordion/AccordionItem";
@import "Button/Button";
@import "Caption/Caption";
@import "GridLayout/GridLayout";
@import "Heading/Heading";
@import "Icon/Icon";
@import "Icon/Icons";
@import "Input/Input";
@import "Paragraph/Paragraph";
@import "Subtitle/Subtitle";
@import "ThemeProvider/ThemeProvider";
@import "Tooltip/Tooltip";