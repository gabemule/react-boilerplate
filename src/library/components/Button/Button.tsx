import React, { HTMLAttributes } from "react";
// import withCustomProps from '@/library/decorators/withCustomProps'
import PropTypes from 'prop-types';

// config
import ButtonConfig from "./Button.config";

// components
import { DsIcon } from "@/library/components/Icon";
// import Button from "@/app/views/components/Button";

export interface DsButtonProps extends HTMLAttributes<HTMLButtonElement> {
  children: string,
  className?: string,
  size?: string,
  disabled?: boolean,
  icon?: string,
  type?: string,
}

const DsButton = ({
  children,
  className,
  onClick,
  size,
  disabled,
  icon,
  type,
}: DsButtonProps) => {
  const buttonSize = size && ButtonConfig.props.size.options.find((value: string) => value === size) ? `${ButtonConfig.class}--${size}` : `${ButtonConfig.class}__huge`;
  const buttonDisabled = disabled && `${ButtonConfig.class}--disabled`;
  const buttonType = type && `${ButtonConfig.class}--${type}`

  return (
    <button
      className={[ButtonConfig.class, buttonSize, buttonType, buttonDisabled, className && className].filter(classes => classes).join(" ")}
      onClick={!disabled ? onClick : () => { }}
      data-testid="button"
    >
      <span className="ds-button__text">
        {children}
        {icon &&
          <span className="ds-button__icon">
            <DsIcon image={icon} color={type === "danger" ? "neutral-00" : ""} />
          </span>
        }
      </span>
    </button>
  );
};

DsButton.propTypes = {
  type: PropTypes.oneOf(ButtonConfig.props.type.options),
  size: PropTypes.oneOf(ButtonConfig.props.size.options),
  disabled: PropTypes.bool,
  icon: PropTypes.oneOf(ButtonConfig.props.icon.options)
};

DsButton.defaultProps = {
  type: ButtonConfig.props.type.default,
  size: ButtonConfig.props.size.default,
  disabled: ButtonConfig.props.disabled.default,
  icon: ButtonConfig.props.icon.default
};

export default DsButton;
