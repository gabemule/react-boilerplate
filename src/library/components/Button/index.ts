import DsButton from "./Button";
import { DsButtonProps }from "./Button";

export { DsButton };
export type { DsButtonProps }
