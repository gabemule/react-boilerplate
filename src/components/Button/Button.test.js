import React from 'react'
import { shallow } from 'enzyme'

// Components
import { Button } from '@/components/Button'

// Constants
const BUTTON_COMPONENT_TEXT = 'Button Component'
const BUTTON_COMPONENT_CLASS = '.storybook-button'
const BUTTON_COMPONENT_PRIMARY = '.storybook-button--primary'
const BUTTON_COMPONENT_SECONDARY = '.storybook-button--secondary'

/**
 * Setup method to avoid code repeat
 * @param props
 * @returns {{wrapper, props}}
 */
function setup(props) {
  const wrapper = shallow(<Button {...props} label={BUTTON_COMPONENT_TEXT} />)
  return { wrapper, props }
}

describe('Button Test Suite', () => {
  it('Should match snapshot', () => {
    expect(Button).toMatchSnapshot()
  })

  it('Should have class', () => {
    const { wrapper } = setup()
    expect(wrapper.find(BUTTON_COMPONENT_CLASS).exists()).toBe(true)
  })

  it('Should have primary class', () => {
    const { wrapper } = setup({ primary: true })
    expect(wrapper.find(BUTTON_COMPONENT_PRIMARY).exists()).toBe(true)
  })

  it('Should have secondary class', () => {
    const { wrapper } = setup({ primary: false })
    expect(wrapper.find(BUTTON_COMPONENT_SECONDARY).exists()).toBe(true)
  })

  it('Should have background color', () => {
    const testProp = '#d2d2d2'
    const { wrapper } = setup({ backgroundColor: testProp })
    expect(
      wrapper.find(BUTTON_COMPONENT_CLASS).props().style.backgroundColor
    ).toBe(testProp)
  })
})
