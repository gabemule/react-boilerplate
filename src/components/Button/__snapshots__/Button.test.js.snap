// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Test Suite Should match snapshot 1`] = `[Function]`;
